### PR TITLE
fix: error pending on socket error

### PR DESCRIPTION
**Pull request by ronag**
Fixes: https://github.com/nodejs/undici/issues/418

### CODE REVIEW

**Comment by ronag** [comment]
@dnlup would you like to try to sort out the tests and also add additional tests (e.g. the case you showed in the [issue](https://github.com/nodejs/undici/issues/418))?

**Comment by dnlup** [comment]
@ronag yes! Is it ok if I do it tomorrow?

**Comment by dnlup** [comment]
After doing some digging, I think the issue is caused by moving the timeout initialization [here](https://github.com/nodejs/undici/blob/v2.0.1/lib/core/request.js#L142) from [here](https://github.com/nodejs/undici/blob/v1.3.1/lib/request.js#L181). I'll work on it.

**Comment by dnlup** [comment]
I think I am going to expose a `onStart` method on the `Request` to allow passing the `resume` function from the `client`.

**Comment by ronag** [comment]
I'm also fine wiht doing it like in this PR, i.e. fail all pending requests on socket error and leave it up to the user to retry. I think that's actually better in some ways. Gives more control to the user.

**Comment by dnlup** [comment]
Ok, I agree. I'll work on the tests then :wink:

File: lib/core/client.js [hunk 1]
```diff
@@ -626,15 +626,7 @@ function onSocketError (err) {
...
       client[kQueue][client[kPendingIdx]++].onError(err)
     }
-  } else if (
-    !client.running &&
-    err.code !== 'ECONNRESET' &&
-    err.code !== 'ECONNREFUSED' &&
-    err.code !== 'EHOSTUNREACH' &&
-    err.code !== 'EHOSTDOWN' &&
-    err.code !== 'UND_ERR_SOCKET' &&
-    err.code !== 'UND_ERR_INFO'
```
Comment (ronag) [inline lib/core/client.js:636]: UND_ERR_INFO should probably still be here

Comment (dnlup) [reply lib/core/client.js:636]: Should the client try to reconnect with those error codes?

**Comment by dnlup** [comment]
I opened https://github.com/nodejs/undici/pull/423 trying to address your suggestions.

**Pull request closed without merging**